simpledb/query/predicate.go

```diff
@@ -42,3 +46,44 @@ func (p *Predicate) String() string {
 	}
 	return strings.Join(terms, " and ")
 }
+
+// 実装都合で Plan のインターフェース相当のものをquery packageに定義
+type planLike interface {
+	Open() (Scan, error)
+	BlocksAccessed() int
+	RecordsOutput() int
+	DistinctValues(fieldName string) int
+	Schema() *record.Schema
```

Comment:
[imo]

ここは、循環参照しないように `Plan` interfaceの置き場所をいい感じに移動するというよりは、

現状通り、疎結合に、termはtermで「こんな感じのinterfaceを満たす引数を要求する」でいいのかもですね

要求（呼び出してる）のは `DistinctValues()` だけっぽいので

```suggestion
// plan.Plan インターフェースの一部を要求する
type planLike interface {
	DistinctValues(fieldName string) int
```

だけの方が設計意図として読みやすいかも

---

`type distinctCountable interface {` とかの方がイケメンかもですがわかりやすいかというと？